overlord/devicestate/devicestate_install_mode_test.go

```diff
@@ -435,6 +438,310 @@ func (s *deviceMgrInstallModeSuite) TestInstallExpTasks(c *C) {
+
+	// we did request a restart through restartSystemToRunModeTask
+	c.Check(s.restartRequests, DeepEquals, []restart.RestartType{restart.RestartSystemNow})
+	c.Check(applyPreseedCalled, Equals, 1)
+}
+
+func (s *deviceMgrInstallModeSuite) TestInstallRestoresPreseedArtifactError(c *C) {
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	restore = devicestate.MockInstallRun(func(mod gadget.Model, gadgetRoot, kernelRoot, device string, options install.Options, _ gadget.ContentObserver, _ timings.Measurer) (*install.InstalledSystemSideData, error) {
+		return nil, nil
+	})
+	defer restore()
+
+	var applyPreseedCalled int
+	restoreApplyPreseed := devicestate.MockMaybeApplyPreseededData(func(st *state.State, ubuntuSeedDir, sysLabel, writableDir string) (bool, error) {
+		applyPreseedCalled++
+		return false, fmt.Errorf("boom")
+	})
+	defer restoreApplyPreseed()
+
+	err := ioutil.WriteFile(filepath.Join(dirs.GlobalRootDir, "/var/lib/snapd/modeenv"),
+		[]byte("mode=install\nrecovery_system=20200105\n"), 0644)
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	s.makeMockInstallModel(c, "dangerous")
+	s.makeMockInstalledPcGadget(c, "", "")
+	devicestate.SetSystemMode(s.mgr, "install")
+	s.state.Unlock()
+
+	s.settle(c)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	installSystem := s.findInstallSystem()
+	c.Check(installSystem.Err(), ErrorMatches, "cannot perform the following tasks:\\n- Ensure next boot to run mode \\(boom\\)")
+
+	c.Check(s.restartRequests, HasLen, 0)
+	c.Check(applyPreseedCalled, Equals, 1)
+}
+
+type fakeSeed struct {
+	modeSnaps      []*seed.Snap
+	essentialSnaps []*seed.Snap
+}
+
+func (fakeSeed) LoadAssertions(db asserts.RODatabase, commitTo func(*asserts.Batch) error) error {
+	return nil
+}
+
+func (fakeSeed) Model() *asserts.Model {
+	return nil
+}
+
+func (fakeSeed) Brand() (*asserts.Account, error) {
+	return nil, nil
+}
+
+func (fakeSeed) LoadEssentialMeta(essentialTypes []snap.Type, tm timings.Measurer) error {
+	return nil
+}
+
+func (fakeSeed) LoadMeta(string, timings.Measurer) error {
+	return nil
+}
+
+func (fakeSeed) UsesSnapdSnap() bool {
+	return true
+}
+
+func (fakeSeed) SetParallelism(n int) {}
+
+func (f *fakeSeed) EssentialSnaps() []*seed.Snap {
+	return f.essentialSnaps
+}
+
+func (f *fakeSeed) ModeSnaps(mode string) ([]*seed.Snap, error) {
+	return f.modeSnaps, nil
+}
+
+func (f *fakeSeed) NumSnaps() int {
+	return 0
+}
+
+func (f *fakeSeed) Iter(func(sn *seed.Snap) error) error {
+	return nil
+}
+
+func (s *deviceMgrInstallModeSuite) mockPreseedAssertion(c *C, model *asserts.Model, sysLabel string, digest string, snaps []interface{}) {
+	headers := map[string]interface{}{
+		"type":              "preseed",
+		"authority-id":      model.AuthorityID(),
+		"series":            "16",
+		"brand-id":          model.BrandID(),
+		"model":             model.Model(),
+		"system-label":      sysLabel,
+		"artifact-sha3-384": digest,
+		"timestamp":         time.Now().UTC().Format(time.RFC3339),
```

Comment:
nitpick: unless the current time is really required, I'd prefer hardcoding a fixed time, to reduce variability in the tests.

Reply:
It is, it's common for all assertion-related tests, otherwise you get an ".... outside of signing key validity" error when the time is too old (or in the future).